app/boot_control/grub.py

```diff
@@ -27,48 +40,458 @@
 )
 
 
-class GrubController(
-    VersionControlMixin, OTAStatusMixin, SlotInUseMixin, BootControllerProtocol
-):
-    def __init__(self) -> None:
+@dataclass
+class GrubMenuEntry:
+    """
+    NOTE: should only be called by the get_entry method
+
+    linux: vmlinuz-<ver>
+    linux_ver: <ver>
+    initrd: initrd.img-<ver>
+    """
+
+    linux_ver: str
+    title: str
+    menuentry: str
+    linux: str
+    initrd: str
+    rootfs_uuid: str
+
+    def __init__(self, ma: re.Match) -> None:
+        """
+        NOTE: check GrubHelper for capturing group definition
+        """
+        self.linux_ver = ma.group("kernel_ver")
+        self.title = ma.group("title")
+        self.menuentry = ma.group("menu_entry")
+
+        # get linux and initrd from menuentry
+        if linux_ma := GrubHelper.linux_pa.search(self.menuentry):
+            self.linux = linux_ma.group("kernel")
+            self.rootfs_uuid = linux_ma.group("rootfs_uuid")
+        if initrd_ma := GrubHelper.initrd_pa.search(self.menuentry):
+            self.initrd = initrd_ma.group("initrd")
+
+        assert self.linux and self.initrd, "failed to detect linux img and initrd"
+        assert self.rootfs_uuid, "failed to detect rootfs uuid"
+
+
+class GrubHelper:
+    menuentry_pa: ClassVar[re.Pattern] = re.compile(
+        # whole capture group
+        r"^(?P<menu_entry>\s*menuentry\s+"
+        # menuentry title:
+        # format: <os>, with Linux <kernel_ver>[ [<recovery>] [<os_probe>]]
+        r"'(?P<title>[^,]*, +with +Linux +(?P<kernel_ver>[^'\s]*)(?P<other>( *\([^\)]*\))*)?)'"
+        r"[^\{]*"  # menuentry options
+        r"\{(?P<entry>[^\}]*)\}"  # menuentry block
+        r")",  # end of whole capture
+        re.MULTILINE | re.DOTALL,
+    )
+    linux_pa: ClassVar[re.Pattern] = re.compile(
+        r"(?P<left>^\s+linux.*(?P<kernel>vmlinuz-(?P<ver>[\.\w\-]*)) +.*)"
+        r"(?P<rootfs>root=UUID=(?P<rootfs_uuid>[\w\-]*))"
+        r"(?P<right>.*)",
+        re.MULTILINE,
+    )
+    initrd_pa: ClassVar[re.Pattern] = re.compile(
+        r"^\s+initrd.*(?P<initrd>initrd.img-(?P<ver>[\.\w-]*))", re.MULTILINE
+    )
+
+    VMLINUZ = "vmlinuz"
+    INITRD = "initrd.img"
+    SUFFIX_OTA = "ota"
+    SUFFIX_OTA_STANDBY = "ota.standby"
+    KERNEL_OTA = f"{VMLINUZ}-{SUFFIX_OTA}"
+    KERNEL_OTA_STANDBY = f"{VMLINUZ}-{SUFFIX_OTA_STANDBY}"
+    INITRD_OTA = f"{INITRD}-{SUFFIX_OTA}"
+    INITRD_OTA_STANDBY = f"{INITRD}-{SUFFIX_OTA_STANDBY}"
+
+    grub_default_options: ClassVar[Dict[str, str]] = {
+        "GRUB_TIMEOUT_STYLE": "menu",
+        "GRUB_TIMEOUT": "10",
+        "GRUB_DISABLE_SUBMENU": "y",
+        "GRUB_DISABLE_OS_PROBER": "true",
+        "GRUB_DEFAULT": "'Ubuntu, with Linux ota'",
```

Comment:
What is this `"'Ubuntu, with Linux ota'"` setting?
In most cases, a number is set for GRUB_DEFAULT.

Reply:
That is a legacy way to specific a boot entry by the submenu title, normal boot entry that use `vmlinuz-ota` to boot on Ubuntu will have `Ubuntu, with Linux ota` as boot menu title.

Here it is just used as a place holder and a last resort, we will always specify a boot entry number when generating grub_cfg.

Reply:
The `GRUB_DEFAULT` option is removed from the `GrubHelper.grub_default_options`, we will not change the `GRUB_DEFAULT` option if default entry index is not provided when calling `update_grub_default` method.